revocation/revocation.go

```diff
@@ -0,0 +1,249 @@
+		resp, err := r.ocspRequest(cert, issuer, server)
+		if err != nil {
+			// If there is a server error, attempt all servers before determining what to return to the user
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+		if time.Now().After(resp.NextUpdate) {
+			err = errors.New("expired OCSP response")
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+		foundNoCheck := false
+		pkixNoCheckOID := "1.3.6.1.5.5.7.48.1.5"
+		for _, extension := range resp.Extensions {
+			if !foundNoCheck && extension.Id.String() == pkixNoCheckOID {
+				foundNoCheck = true
+			}
+		}
+		if !foundNoCheck {
+			// This will be ignored until CRL is implemented
+			// If it isn't found, CRL should be used to verify the OCSP response
+			fmt.Printf("\n[WARNING] An ocsp signing cert is missing the id-pkix-ocsp-nocheck extension (%s)\n", pkixNoCheckOID)
+		}
+
+		if resp.Status == ocsp.Revoked {
+			if time.Now().After(resp.RevokedAt) {
+				(*results)[resultIndex] = RevokedInOCSPError{}
+				return RevokedInOCSPError{}
+			} else {
+				(*results)[resultIndex] = nil
+				return nil
+			}
+		} else if resp.Status == ocsp.Good {
+			(*results)[resultIndex] = nil
+			return nil
+		} else {
+			(*results)[resultIndex] = UnknownInOCSPError{}
+			return UnknownInOCSPError{}
+		}
+
+	}
+	// Errors in all server responses, determine the most pressing
+	err = r.mergeErrorsFunc(serverErrs)
+	(*results)[resultIndex] = err
+	return err
+}
+
+func (r *revocation) ocspRequest(cert, issuer *x509.Certificate, server string) (*ocsp.Response, error) {
+	ocspRequest, err := ocsp.CreateRequest(cert, issuer, &ocsp.RequestOptions{Hash: crypto.SHA1})
+	if err != nil {
+		return nil, err
+	}
+
+	var resp *http.Response
+	if len(ocspRequest) > 256 {
+		buf := bytes.NewBuffer(ocspRequest)
+		resp, err = r.httpClient.Post(server, "application/ocsp-request", buf)
+	} else {
+		reqURL := server + "/" + url.QueryEscape(base64.StdEncoding.EncodeToString(ocspRequest))
+		resp, err = r.httpClient.Get(reqURL)
+	}
+
+	if err != nil {
+		if urlErr, ok := err.(*url.Error); ok && urlErr.Timeout() {
+			return nil, OCSPTimeoutError{}
+		}
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("failed to retrieve OSCP: response had status code %d", resp.StatusCode)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case bytes.Equal(body, ocsp.UnauthorizedErrorResponse):
+		return nil, errors.New("OSCP unauthorized")
+	case bytes.Equal(body, ocsp.MalformedRequestErrorResponse):
+		return nil, errors.New("OSCP malformed")
+	case bytes.Equal(body, ocsp.InternalErrorErrorResponse):
+		return nil, errors.New("OSCP internal error")
+	case bytes.Equal(body, ocsp.TryLaterErrorResponse):
+		return nil, errors.New("OSCP try later")
+	case bytes.Equal(body, ocsp.SigRequredErrorResponse):
+		return nil, errors.New("OSCP signature required")
+	}
```

Comment:
should we automatically retry for some of this errors?

Reply:
@priteshbandi does this explanation for not having a retry answer your question? If so, I can resolve this